enhancements/machine-health-checking.md

```diff
@@ -0,0 +1,204 @@
+---
+title: machine-health-checking
+authors:
+  - @bison
+  - @ingvagabund
+  - @enxebre
+  - @cynepco3hahue
+reviewers:
+  - @derekwaynecarr
+  - @michaelgugino
+  - @beekhof
+  - @mrunalp
+approvers:
+  - @enxebre
+  - @bison
+  - @derekwaynecarr
+  - @michaelgugino
+
+creation-date: 2019-09-09
+last-updated: 2019-09-09
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Machine health checking
+
+## Release Signoff Checklist
+
+ - [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs
+
+
+## Summary
+
+Enable opt in for automated health checking and remediation of unhealhty nodes backed by machines.
+
+## Motivation
+
+Reduce administrative overhead to run a cluster and ability to respond to failures of machines.
+
+### Goals
+
+- Providing a default remediation strategy for public clouds - Machine deletion.
+
+- Allowing to define different unhealthy criterias for different pools of machines.
+
+- Allowing new and custom remediation strategies developed outside the machine-healthcheck controller, e.g baremetal reboot.
+
+- Providing administrator-defined short-circuiting of automated remediation.
+
+
+### Non-Goals
+
+- Coordinating individual remediation systems. This must be driven by each system however they see fit, e.g. by annotating affected nodes with state and transition time.
+
+- Recording long-term stable history of all health-check failures or remediations.
+
+## Proposal
+
+- As a customer I only care about my apps availability so I want my cluster infrastructure to be self healing and the nodes to be remediated automatically.
+
+- As a cluster admin I want to define my unhealthy criteria as a list of node conditions https://github.com/kubernetes/node-problem-detector.
+
+- As a cluster admin I want to define different unhealthy criteria for pools of machines targeted for different workloads.
```

Comment:
I disagree with this point.  A node is unhealthy, or it isn't.  Having to support different healthy criteria across arbitrary groups (of which, we haven't defined what a 'group' is) adds additional complexity to development, use, and support.

Reply:
A group here is the group of machines that matches the label selector. An administrator may want more or less aggressive remediation for some machines. That would be expressed in this model by multiple `MachineHealthCheck` resources with different label selectors and different `unhealthyConditions`, with e.g. a higher timeout value.

Reply:
An unhealthy criteria is just a list of nodeConditions https://github.com/openshift/enhancements/pull/9/files#diff-53036ce0d8589cdbed15ed2e95076b54R101
I can't see how that can add any complexity. Allowing to define your criteria as a list of conditions grants awareness and visibility of problems and enable the MHC to benefit from problem detector tooling like https://github.com/kubernetes/node-problem-detector

Reply:
How about the case of a large ceph node, where a single disk error is certainly not a reason to flush the whole node?